API provides search functionality for drinks

Carltesio opened the pull request:
# [PT](https://www.pivotaltracker.com/story/show/171703982)

## Changes proposed in this pull request:
* Adds unit test for Cocktails controller
* Adds unit test for 200/400 error in case of empty search
* Sanitized search to get only name, ID, category and IBA key from API database

## What I have learned working on this feature:
* Setup configuration of stub in the test environment for an empty body on the API request for drinks.
* Sanitizing data search array from API database. 


## Gems installed:
* gem 'rest-client'
* gem 'webmock'

Carltesio committed:
adding first test (response200)

janko101 committed:
Creates Cocktails Controller + configures routes for cocktails controller

janko101 committed:
Installs gem rest-client + defining the index response in the Cocktails Controller

Commit:
adds search tests

Commit:
modifies cocktails controller and adds empty query test

Carltesio committed:
adding webmock, fixture file to run test

Commit:
adds aditional stub in controller

Commit:
modifies rails helper and the controller

kaylawoodbury commented:
Sorry, I was on the wrong page and accidentally assigned the reviewer, but removed it so it should be fine now.

In the PR details, make sure you include a section for the gems you are adding in the PR. You can also remove the #1.2 from the header. We should probably remove those from the PT as well based on our conversation with Thomas yesterday.

kaylawoodbury commented on Gemfile:
```diff
@@ -8,8 +8,10 @@ gem 'pg', '>= 0.18', '< 2.0'
 gem 'puma', '~> 4.1'
 gem 'bootsnap', '>= 1.4.2', require: false
 gem 'rack-cors', require: 'rack/cors'
+gem 'rest-client'
```
Since gems are being added be sure to include them in the PR description. The format is pinned in the group channel I think.

Carltesio replied:
added the gems to the PR

kaylawoodbury commented on app/controllers/api/v1/cocktails_controller.rb:
```diff
@@ -0,0 +1,32 @@
+class Api::V1::CocktailsController < ApplicationController
+    before_action :check_query_param
```
Make sure that a consistent tab formatting is being used, it looks like the gemfile is a 2 space tab whereas this file is a 4 space tab. I think option+shift+f or alt+shift+f can be used for auto-formatting.

Carltesio replied:
Yes they were 4 spaces. Took it down to 2.

kaylawoodbury commented on app/controllers/api/v1/cocktails_controller.rb:
```diff
@@ -0,0 +1,32 @@
+            "https://www.thecocktaildb.com/api/json/v1/1/search.php",
+            {
+                params: {
+                    s: params[:q]
```
I don't know if this is out of scope for this PR, but it doesn't seem like we are sanitizing the data but pulling back all data for the drinks in the results. Is this something that would be relevant to include in this PR?

Carltesio replied:
Fixed it with Aditya

kaylawoodbury reviewed (commented):
I don't see any errors per se based on my knowledge level, but I am wondering if there are a few more steps/items that would be applicable in this PR. Primarily, if we need to sanitize the data on the backend so that we will only be pushing the relevant data to the client?

janko101 committed:
Trying to sanitaze result data for cocktail search

janko101 committed:
Sanitazing search cocktails data to show only name and id

Carltesio committed:
formatting code for PR as request from Kayla

kianaditya commented:
Fix PT link

kianaditya commented on config/routes.rb:
```diff
@@ -1,2 +1,7 @@
 Rails.application.routes.draw do
+    namespace :api do
+        namespace :v1 do
+            resources :cocktails, only:[:index]
```
Please fix formatting

kianaditya commented on app/controllers/api/v1/cocktails_controller.rb:
```diff
@@ -0,0 +1,39 @@
+
+  def check_query_param
+      if params[:q] == ""
+          render json: {error: 'No input in search field'}, status: 400
```
Please fix formatting

tochman commented on app/controllers/api/v1/cocktails_controller.rb:
```diff
@@ -0,0 +1,39 @@
+              {
+                  "name": drinks["drinks"][0]["strDrink"],
+                  "id": drinks["drinks"][0]["idDrink"]
+              }
```
Why are you only returning one drink in the index action?

DanielGITB replied:
We realise that now and will try to address it.

kianaditya commented on app/controllers/api/v1/cocktails_controller.rb:
```diff
@@ -0,0 +1,39 @@
+  private
+
+  def check_query_param
+      if params[:q] == ""
```
Ideally you should handle this in the client so that there is no unnecessary network call

DanielGITB replied:
Thats good to know. For consistency we will leave it here as we worked through this with Oliver

tochman commented on e(response.body:
```diff
@@ -0,0 +1,272 @@
+   [31m[1;31m"[0m[31mstrMeasure14[1;31m"[0m[31m[0m=>[1;36mnil[0m,
+   [31m[1;31m"[0m[31mstrMeasure15[1;31m"[0m[31m[0m=>[1;36mnil[0m,
+   [31m[1;31m"[0m[31mstrCreativeCommonsConfirmed[1;31m"[0m[31m[0m=>[31m[1;31m"[0m[31mNo[1;31m"[0m[31m[0m,
+   [31m[1;31m"[0m[31mdateModified[1;31m"[0m[31m[0m=>[31m[1;31m"[0m[31m2015-08-18 14:41:51[1;31m"[0m[31m[0m}]]
```
??

tochman commented on quit:
```diff
@@ -0,0 +1,271 @@
+  [31m[1;31m"[0m[31mstrMeasure13[1;31m"[0m[31m[0m=>[1;36mnil[0m,
+  [31m[1;31m"[0m[31mstrMeasure14[1;31m"[0m[31m[0m=>[1;36mnil[0m,
+  [31m[1;31m"[0m[31mstrMeasure15[1;31m"[0m[31m[0m=>[1;36mnil[0m,
+  [31m[1;31m"[0m[31mstrCreativeCommonsConfirmed[1;31m"[0m[31m[0m=>[31m[1;31m"[0m[31mNo[1;31m"[0m[31m[0m,
```
???

DanielGITB replied:
We deleted this and the quit-file

kianaditya commented on app/controllers/api/v1/cocktails_controller.rb:
```diff
@@ -0,0 +1,39 @@
+  before_action :check_query_param
+
+  def index
+      response = RestClient.get(
```
Something to think about - What if you don't get ANY response from the api ??

DanielGITB replied:
Do we need to add a 400 error message if we don't get any response from the api?

kianaditya commented on spec/requests/api/v1/user_can_search_cocktail_by_name_spec.rb:
```diff
@@ -0,0 +1,52 @@
+            expect(response.status).to eq 200
+        end
+
+        it 'returns drink' do
```
Are you returning only one drink ? also check for count

DanielGITB replied:
We can add a count test

janko101 committed:
Adds sanitized search data method

Commit:
resolving merge issue

Carltesio committed:
solving merging problems

Carltesio committed:
Fixed merging issue by accepting incoming changes
Merge branch 'search_by_name' of github.com:DanielGITB/cocktails_api_team_1 into search_by_name

janko101 committed:
Fixing issues

janko101 committed:
Trying to resolve merge issues

Commit:
resloving pull issue

Commit:
Merge branch 'search_by_name' of https://github.com/MadFarmer101/cocktails_api_team_1 into search_by_name
resolving pull issue

Carltesio committed:
fixing merge conflict

Commit:
adds funcitonal code to cocktailss_controller

Carltesio committed:
Merge branch 'search_by_name' of github.com:DanielGITB/cocktails_api_team_1 into search_by_name

Commit:
adds id and amount of drinks tests

Carltesio committed:
Merge branch 'search_by_name' of github.com:DanielGITB/cocktails_api_team_1 into search_by_name

Carltesio committed:
adding test abi and catgory2

Carltesio committed:
adding CAtegory and IBA from API database to pass the tests

Carltesio committed:
formatting codde

Carltesio committed:
formatting code and cancelling quit file

kianaditya commented on spec/requests/api/v1/user_can_search_cocktail_by_name_spec.rb:
```diff
@@ -0,0 +1,69 @@
+describe 'GET /cocktails', type: :request do
```
Please fix formatting for this file

Carltesio committed:
formatting Spec file

Carltesio committed:
formatting Spec file and rails helper

kianaditya reviewed (approved):
LGTM
